### PR TITLE
feat(neo4j): added support to Neo4j operations

### PR DESCRIPTION
**Pull request by osherv**

### CODE REVIEW

File: package-lock.json [hunk 1]
```diff
@@ -2364,6 +2364,14 @@
       "integrity": "sha1-YXmX/F9gV2iUxDX5QNgZ4TW4B2I=",
       "dev": true
     },
+    "ansi-align": {
```
Comment (haimrait) [inline package-lock.json]: Why did you change the package-lock.json
Did you change the package.json?

File: src/events/neo4j.js [hunk 2]
```diff
@@ -0,0 +1,265 @@
...
+const tracer = require('../tracer.js');
+const errorCode = require('../proto/error_code_pb.js');
+
+/**
+ * Extracts the relevant arguments from provided arguments
+ * @returns {Object} Object of the extracted arguments
+ */
+function getArgsFromFunction(...args) {
+    return {
+        query: args[0],
```
Comment (haimrait) [inline src/events/neo4j.js]: what about transactionConfig?
https://neo4j.com/docs/api/javascript-driver/4.2/class/src/session.js~Session.html

File: src/events/neo4j.js [hunk 3]
```diff
@@ -0,0 +1,265 @@
...
+    return { metadata, fullMetadata };
+}
+
+
+/**
+ * Wraps neo4j.Session.run and neo4j.Transaction.run functions with tracing
+ * @param {Function} wrappedFunction The function to wrap from Neo4j
+ * @returns {Function} The wrapped function
+ */
+function neo4jSessionRunWrapper(wrappedFunction) {
```
Comment (haimrait) [inline src/events/neo4j.js:271]: Maybe spread it to two different wrappers that reuse the same code for better readability
like (neo4jSessionRunWrapper, neo4jTranssactionRunWrapper)

File: src/events/neo4j.js [hunk 4]
```diff
@@ -0,0 +1,265 @@
...
+ * Parse from DB session with Neo4j server, useful metadata
+ * @param {neo4j.Session|neo4j.Transaction} session A DB session (could be Session of Transaction)
+ * @returns {{metadata: Object, fullMetadata: Object}} JSONs with command's metadata and command's
+ *      full metadata which contains additional data about the event
+ */
+function getNeo4JSessionMetadata(session) {
+    const metadata = {};
+    const fullMetadata = {};
+
+    let connectionHolder;
```
Comment (haimrait) [inline src/events/neo4j.js:98]: Definition is not enough it will raise an exception if you don't init this value in case of default

File: src/events/neo4j.js [hunk 5]
```diff
@@ -0,0 +1,265 @@
...
+        break;
+
+    default:
+        break;
+    }
+
+    metadata.dbName = connectionHolder._database ? connectionHolder._database : 'default';
+    metadata.operation = `${session.constructor.name}:${operationMode}`;
+
+    fullMetadata.config = connectionHolder._connectionProvider._config;
```
Comment (haimrait) [inline src/events/neo4j.js]: add safe checks

File: src/events/neo4j.js [hunk 6]
```diff
@@ -0,0 +1,265 @@
...
+}
+
+
+/**
+ * Gets session's address info from Transaction/Session instance
+ * @param {neo4j.Session|neo4j.Transaction} session A DB session (could be Session of Transaction)
+ * @returns {Object} json with address info { host, port }
+ */
+function getAddressInfo(session) {
+    let port = '7474';
```
Comment (haimrait) [inline src/events/neo4j.js]: change host and port to consts

File: src/events/neo4j.js [hunk 7]
```diff
@@ -0,0 +1,265 @@
...
+                'neo4j',
+                0,
+                errorCode.ErrorCode.OK,
+            ]);
+
+            dbApiEvent.setResource(resource);
+
+            eventInterface.addToMetadata(dbApiEvent, metadata, fullMetadata);
+
+            const resultResponse = wrappedFunction.apply(this, [query, params]);
```
Comment (haimrait) [inline src/events/neo4j.js]: what about transactionConfig?

Comment (haimrait) [reply src/events/neo4j.js]: https://neo4j.com/docs/api/javascript-driver/4.2/class/src/session.js~Session.html

File: src/events/neo4j.js [hunk 8]
```diff
@@ -0,0 +1,265 @@
...
+                    // eslint-disable-next-line require-jsdoc
+                    onNext(record) {
+                        records.push(record);
+                        return observer.onNext.apply(this, arguments);
+                    },
+
+                    // eslint-disable-next-line require-jsdoc
+                    onCompleted(summary) {
+                        switch (summary.queryType) {
+                        case 's':
```
Comment (haimrait) [inline src/events/neo4j.js]: why s is Read?

File: src/events/neo4j.js [hunk 9]
```diff
@@ -0,0 +1,265 @@
...
+                        // eslint-disable-next-line prefer-rest-params
+                        return observer.onError.apply(this, arguments);
+                    },
+                });
+            };
+
+            return resultResponse;
+        } catch (error) {
+            tracer.addException(error);
+            return wrappedFunction.apply(this, [query, params]);
```
Comment (haimrait) [inline src/events/neo4j.js]: what about transactionConfig?
https://neo4j.com/docs/api/javascript-driver/4.2/class/src/session.js~Session.html

File: src/events/neo4j.js [hunk 10]
```diff
@@ -0,0 +1,265 @@
...
+
+            // Override the Result subscribe with patched subscriber
+            // eslint-disable-next-line func-names
+            resultResponse.subscribe = function (observer) {
+                const records = [];
+
+                return originalSubscribe.call(this, {
+                    ...observer,
+                    // eslint-disable-next-line no-unused-vars,require-jsdoc
+                    onKeys(_keys) {
```
Comment (haimrait) [inline src/events/neo4j.js]: I think onKeys useless

File: src/events/neo4j.js (hunk 6 above)
Comment (osherv) [reply src/events/neo4j.js]: They can not be, they might be changed later this function.
should i change the function design?

File: src/events/neo4j.js (hunk 4 above)
Comment (osherv) [reply src/events/neo4j.js:98]: Fixed by added later this function 'if(connectionHandler)'
Is that the right way?
Otherwise i need to init a new connectionHandler instance, right?

File: src/events/neo4j.js (hunk 5 above)
Comment (osherv) [reply src/events/neo4j.js]: Fixed? (does that fix also fixed the prev comment?)

File: src/events/neo4j.js (hunk 6 above)
Comment (haimrait) [reply src/events/neo4j.js]: I meant change 7474 and neo4j to costs and not exact values

Comment (osherv) [reply src/events/neo4j.js]: oh oh, fixed :O

File: src/events/neo4j.js (hunk 7 above)
Comment (osherv) [reply src/events/neo4j.js]: fixed! ty

File: src/events/neo4j.js (hunk 8 above)
Comment (osherv) [reply src/events/neo4j.js]: Fixed!

File: src/events/neo4j.js (hunk 10 above)
Comment (osherv) [reply src/events/neo4j.js]: Fixed!

File: src/events/neo4j.js (hunk 3 above)
Comment (osherv) [reply src/events/neo4j.js:271]: Finally Fixed XD

File: src/events/neo4j.js (hunk 2 above)
Comment (osherv) [reply src/events/neo4j.js]: Fixed

File: src/events/neo4j.js (hunk 9 above)
Comment (osherv) [reply src/events/neo4j.js]: Fixed

File: src/events/neo4j.js [hunk 11]
```diff
@@ -0,0 +1,386 @@
...
+                    arguments,
+                    dbApiEvent,
+                    startTime
+                ));
+            };
+
+            return resultResponse;
+        } catch (error) {
+            tracer.addException(error);
+            return wrappedFunction.apply(this, [query, params]);
```
Comment (haimrait) [inline src/events/neo4j.js]: I think if originalSubscribe.call will raise an exception we will call 
wrappedFunction.apply twice the first one before the subscribe function and the second inside the catch maybe we need to change the place of the catch scope

File: src/events/neo4j.js [hunk 12]
```diff
@@ -0,0 +1,386 @@
...
+                    arguments,
+                    dbApiEvent,
+                    startTime
+                ));
+            };
+
+            return resultResponse;
+        } catch (error) {
+            tracer.addException(error);
+            return wrappedFunction.apply(this, [query, params, transactionConfig]);
```
Comment (haimrait) [inline src/events/neo4j.js]: same as below

File: src/events/neo4j.js (hunk 4 above)
Comment (haimrait) [reply src/events/neo4j.js:98]: yea its ok

File: src/events/neo4j.js [hunk 13]
```diff
@@ -0,0 +1,386 @@
...
+    session,
+    observer,
+    sessionArguments,
+    dbApiEvent,
+    startTime
+) {
+    const records = [];
+
+    return {
+        // eslint-disable-next-line require-jsdoc
```
Comment (haimrait) [inline src/events/neo4j.js:169]: what about the spread of ...observer
why did you remove it?

**Review by haimrait — CHANGES_REQUESTED** [review]

File: src/events/neo4j.js (hunk 13 above)
Comment (osherv) [reply src/events/neo4j.js:169]: By mistake, fixed, ty!

File: src/events/neo4j.js (hunk 12 above)
Comment (osherv) [reply src/events/neo4j.js]: Fixed by moving the apply out of the try scope, i think that will solve the issue, right?

File: src/events/neo4j.js (hunk 11 above)
Comment (osherv) [reply src/events/neo4j.js]: Fixed by moving the apply out of the try scope, i think that will solve the issue, right?

File: src/events/neo4j.js [hunk 14]
```diff
@@ -0,0 +1,387 @@
...
+function neo4jTransactionRunWrapper(wrappedFunction) {
+    return function internalNeo4jTransactionRunWrapper(...args) {
+        const relevantArgs = getArgsFromFunction(...args, this.constructor.name);
+        const {
+            query, params,
+        } = relevantArgs;
+
+        utils.debugLog('User called Neo4j wrapped Transaction run function');
+
+        const resultResponse = wrappedFunction.apply(this, [query, params]);
```
Comment (haimrait) [inline src/events/neo4j.js]: Now if the customer's code throws an exception we will not add it to the tracer object

File: src/events/neo4j.js [hunk 15]
```diff
@@ -0,0 +1,387 @@
...
+function neo4jSessionRunWrapper(wrappedFunction) {
+    return function internalNeo4jSessionRunWrapper(...args) {
+        const relevantArgs = getArgsFromFunction(...args, this.constructor.name);
+        const {
+            query, params, transactionConfig,
+        } = relevantArgs;
+
+        utils.debugLog('User called Neo4j wrapped Session run function');
+
+        const resultResponse = wrappedFunction.apply(
```
Comment (haimrait) [inline src/events/neo4j.js]: Now if the customer's code throws an exception we will not add it to the tracer object

File: src/events/neo4j.js [hunk 16]
```diff
@@ -0,0 +1,387 @@
...
+        connectionHolder = session._connectionHolder;
+        break;
+
+    default:
+        return {
+            host, port,
+        };
+    }
+
+    if (session && connectionHolder && connectionHolder._connectionProvider) {
```
Comment (haimrait) [inline src/events/neo4j.js]: why do you check the session here?

File: src/events/neo4j.js (hunk 15 above)
Comment (osherv) [reply src/events/neo4j.js]: fixed

File: src/events/neo4j.js (hunk 14 above)
Comment (osherv) [reply src/events/neo4j.js]: fixed

File: src/events/neo4j.js (hunk 16 above)
Comment (osherv) [reply src/events/neo4j.js]: right, fixed. thanks

**Review by haimrait — APPROVED** [review]

**Comment by osherv** [comment]
bors merge
